crates/codegen/parser/runtime/src/query/engine.rs

```diff
@@ -0,0 +1,481 @@
+use std::collections::HashMap;
+use std::rc::Rc;
+
+// This crate is copied to another crate, so all imports should be relative
+use super::super::cst;
+use super::super::cursor::Cursor;
+use super::model::{
+    AlternativesMatcher, BindingMatcher, Kind, Matcher, NodeMatcher, NodeSelector,
+    OneOrMoreMatcher, OptionalMatcher, Query, SequenceMatcher,
+};
+
+impl Cursor {
+    pub fn query(self, queries: Vec<Query>) -> QueryResultIterator {
+        QueryResultIterator::new(self, queries)
+    }
+
+    fn consume(&mut self) -> bool {
+        if self.is_completed() {
+            false
+        } else {
+            if !self.go_to_next_sibling() {
+                self.complete();
+            }
+            true
+        }
+    }
+
+    fn matches_node_selector(&self, node_selectot: &NodeSelector) -> bool {
+        match self.node() {
```

Comment:
For completely private methods like these, I think they can go on whatever type is most regular for the problem being solved. I was in two minds about this, but came down on the cursor being the right place for the sense of the algorithm.

Reply:
As for the `query` method itself, queries are layered on top of cursors - they add to the functionality but don't belong on the 'canonical' definition of the type. I didn't feel the need to make a `Queryable` trait, but maybe that would make sense, and then it could be on `Node` as well. WDYT?

Reply:
Since that Rust allows defining extension methods on either types (modules), I think it is more about the UX when importing and using a `slang::cursor::Cursor` .. if they want to call a `Cursor::query()` method:

- if there is a single possible query that can be returned, and it is not one of many implementations/variations of what a `Query` is .. then I think it should live on the `Cursor` type, because it is a unique API bound only to that type, even if the cursor can be used without the query .. they don't have to import the query module, or know about its types ..
- On the other hand, if we make a `Queryable` trait, then the trait would naturaly live here in the query module, and its implementations would live in the cursor module (and also the `cst` module as you suggested).

Makes sense that a node would be queryable .. if the extra implementation is not too much work, I would say let's do it.